tests/zfs-tests/tests/functional/cache/cache_010_pos.ksh

```diff
@@ -50,51 +50,55 @@ function cleanup_testenv
 	if [[ -n $lofidev ]]; then
 		if is_linux; then
 			losetup -d $lofidev
+		elif is_freebsd; then
+			mdconfig -du ${lofidev#md}
 		else
 			lofiadm -d $lofidev
 		fi
 	fi
 }
 
-log_assert "Cache device can only be block devices."
+log_assert "Verify cache devices can be disk, file, lofi device or any " \
+  "device that presents a block interface"
+
+verify_disk_count "$DISKS" 2
 log_onexit cleanup_testenv
 
 TESTVOL=testvol1$$
 dsk1=${DISKS%% *}
 log_must zpool create $TESTPOOL ${DISKS#$dsk1}
 
-# Add nomal ${DEV_RDSKDIR} device
+# Add normal ${DEV_RDSKDIR} device
 log_must zpool add $TESTPOOL cache \
-    ${DEV_RDSKDIR}/${dsk1}${SLICE_PREFIX}${SLICE0}
+    ${DEV_RDSKDIR}/${dsk1}
+log_must zpool remove $TESTPOOL ${DEV_RDSKDIR}/${dsk1}
+
+
+# Add provided disk
+log_must zpool add $TESTPOOL cache $dsk1
 log_must verify_cache_device $TESTPOOL $dsk1 'ONLINE'
+log_must zpool remove $TESTPOOL $dsk1
 
 # Add normal file
-log_mustnot zpool add $TESTPOOL cache $VDEV2
+log_must zpool add $TESTPOOL cache $LDEV
+ldev=$(random_get $LDEV)
+log_must verify_cache_device $TESTPOOL $ldev 'ONLINE'
 
-# Add /dev/rlofi device (allowed under Linux)
+# Add loop back device
 if is_linux; then
 	lofidev=$(losetup -f)
-	lofidev=${lofidev##*/}
 	log_must losetup $lofidev ${VDEV2%% *}
-	log_must zpool add $TESTPOOL cache $lofidev
-	log_must zpool remove $TESTPOOL $lofidev
-	log_must losetup -d $lofidev
-	lofidev=""
+	lofidev=${lofidev##*/}
+elif is_freebsd; then
+	lofidev=$(mdconfig -a ${LDEV2%% *})
```

Comment:
This (and I assume below for illumos as well) should be `VDEV2` instead. In cache.cfg `LDEV2` is one of the `DISKS` which by default is already a loop back device, so this fails. Using `VDEV2` will give us a file.

With that change this passes on FreeBSD 💯

Reply:
Thank you, I wasn't sure about FreeBSD and Illumos (and in hindsight I should've atleast spun up an image to test with, my bad). Fixed.